### PR TITLE
FIx checked DAC layout issues

### PR DESCRIPTION
**Pull request by sdmaclea**
~~Fix DBI type layout issues~~

+ ~~Add utilcode_dbi~~
+ ~~Consistently define DBI features~~
+ ~~Define T_CRITICAL_SECTION during non-DAC builds~~
+ ~~Fix DBI CMiniMdTemplate layout~~

Fix DAC layout in checked builds

### CODE REVIEW

**Comment by sdmaclea** [comment]
`FEATURE_METADATA_RELEASE_MEMORY_ON_REOPEN` was enabled for parts of the DBI build, but not for all DBI libraries.  This led to multiple definitions of several types, because this enables new members in these types.

File: src/coreclr/src/utilcode/CMakeLists.txt [hunk 1]
```diff
@@ -119,6 +126,14 @@ endif(CLR_CMAKE_HOST_WIN32)
 set_target_properties(utilcode_dac PROPERTIES DAC_COMPONENT TRUE)
 set_target_properties(utilcode_crossgen PROPERTIES CROSSGEN_COMPONENT TRUE)
 target_compile_definitions(utilcode_dac PRIVATE SELF_NO_HOST)
+target_compile_definitions(utilcode_dbi PRIVATE FEATURE_METADATA_EMIT)
```
Comment (jkotas) [inline src/coreclr/src/utilcode/CMakeLists.txt:129]: Where does utilcode depend on any of this?

Comment (sdmaclea) [reply src/coreclr/src/utilcode/CMakeLists.txt:129]: I'm not completely sure, but it is including the `src/coreclr/src/md/inc/metamodel.h` header.

Comment (sdmaclea) [reply src/coreclr/src/utilcode/CMakeLists.txt:129]: I can look into it more carefully.

Comment (jkotas) [reply src/coreclr/src/utilcode/CMakeLists.txt:129]: I guess it is including it to use it in pointer definitions. Those can be just forward decls e.g. `class CMiniMdSchema;`, without including the full header.

Comment (sdmaclea) [reply src/coreclr/src/utilcode/CMakeLists.txt:129]: I am guessing all these need to be fixed

```
git grep "#include .*\.\." -- src/coreclr/src/inc

src/coreclr/src/inc/eventtracebase.h:#include "../gc/env/etmdummy.h"
src/coreclr/src/inc/formattype.h:#include "../ildasm/dynamicarray.h"
src/coreclr/src/inc/metadata.h:#include "../md/inc/metamodelro.h"
src/coreclr/src/inc/metadata.h:#include "../md/inc/liteweightstgdb.h"
src/coreclr/src/inc/metadata.h:#include "../md/inc/metamodel.h"
src/coreclr/src/inc/stgpool.h:#include "../md/hotdata/hotheap.h"
src/coreclr/src/inc/stgpool.h:#include "../md/debug_metadata.h"
```

Comment (jkotas) [reply src/coreclr/src/utilcode/CMakeLists.txt:129]: Yes, it would be nice to not have dependencies from inc to other directories.

Comment (sdmaclea) [reply src/coreclr/src/utilcode/CMakeLists.txt]: I have created a separate change for part of this.  I will create issues to discuss the others.

**Comment by sdmaclea** [comment]
Ping.  @hoyosjs @jkotas This is ready to merge.

**Review by hoyosjs — APPROVED** [review]
Just out of curiosity, what platforms were we seeing the alignment issues in?

**Comment by sdmaclea** [comment]
The alignment issues only occurred in checked builds.
I saw the issue while testing 3.1 backport.  I was using the CI built bits and so I had checked to test for a few platforms.

**Comment by sdmaclea** [comment]
To be specific I think it occurred in all platforms.

**Pull request merged**